05-structs.Rmd

```diff
@@ -0,0 +1,467 @@
+
+To define a struct, we must declare it outside of a `workflow` and `task`:
+
+```         
+struct referenceGenome {
+    File ref_fasta
+    File ref_fasta_index
+    File ref_dict
+    File ref_amb
+    File ref_ann
+    File ref_bwt
+    File ref_pac
+    File ref_sa
+    String ref_name
+}
+
+workflow minidata_mutation_calling_v1 {
+  input {
+    File sampleFastq
+    referenceGenome refGenome
+    ...
+  }
+  # Map reads to reference
+  call BwaMem {
+    input:
+      input_fastq = sampleFastq,
+      refGenome = refGenome
+  }
+}
+```
+
+The `referenceGenome` struct contains all the variables related to the reference genome, but values cannot be defined here. The struct definition merely lays the skeleton components of the data structure, but contains no actual values.
+
+In our workflow inputs, we declare a `referenceGenome` struct variable called `refGenome` via `referenceGenome refGenome` and removed our original reference genome definitions. We can access the variables within a struct by the following syntax: `structVar.varName`, such as `refGenome.ref_name`. The [WDL syntax specs](https://github.com/openwdl/wdl/blob/main/versions/1.0/SPEC.md#struct-definition) has more information on how to define and use structs.
+
+To give values to `refGenome`, we need to modify our JSON metadata file. We define `refGenome` variable in a nested structure that corresponds to the `referenceGenome` struct. Let's take a look:
+
+```         
+{
+  "minidata_mutation_calling_v1.refGenome": {
+    "ref_fasta": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/Homo_sapiens_assembly19.fasta",
+    "ref_fasta_index": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/Homo_sapiens_assembly19.fasta.fai",
+    "ref_dict": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/Homo_sapiens_assembly19.dict",
+    "ref_pac": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/Homo_sapiens_assembly19.fasta.pac",
+    "ref_sa": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/Homo_sapiens_assembly19.fasta.sa",
+    "ref_amb": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/Homo_sapiens_assembly19.fasta.amb",
+    "ref_ann": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/Homo_sapiens_assembly19.fasta.ann",
+    "ref_bwt": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/Homo_sapiens_assembly19.fasta.bwt",
+    "ref_name": "hg19"
+  },
+  "minidata_mutation_calling_v1.dbSNP_vcf_index": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/dbsnp_138.b37.vcf.gz.tbi",
+  ...
+}
+```
+
+Now `refGenome` has all the values it needs for our tasks.
+
+In addition, we have replaced all the reference genome inputs in `call BwaMem` with `refGenome`: we can pass information to a task via structs.
```

Comment:
Consider change to "...in `call BwaMem` with `refGenome` in order to pass information to a task via structs."